api/inc/unsupported.h

```diff
@@ -133,63 +134,64 @@ static inline UVISOR_FORCEINLINE void uvisor_write(uint32_t addr, uint32_t val,
 
-/* The conditional statement will be optimised away since the compiler already
- * knows the sizeof(type). */
-#define ADDRESS_READ(type, addr) \
-    (sizeof(type) == 4 ? uvisor_read32((volatile uint32_t *) (addr)) : \
-     sizeof(type) == 2 ? uvisor_read16((volatile uint16_t *) (addr)) : \
-     sizeof(type) == 1 ? uvisor_read8((volatile uint8_t *) (addr)) : 0)
-
-/* The switch statement will be optimised away since the compiler already knows
- * the sizeof(type). */
-#define ADDRESS_WRITE(type, addr, val) \
-    { \
-        switch(sizeof(type)) \
-        { \
-            case 4: \
-                uvisor_write32((volatile uint32_t *) (addr), (uint32_t) (val)); \
-                break; \
-            case 2: \
-                uvisor_write16((volatile uint16_t *) (addr), (uint16_t) (val)); \
-                break; \
-            case 1: \
-                uvisor_write8((volatile uint8_t *) (addr), (uint8_t) (val)); \
-                break; \
-        } \
-    }
-
-#define UNION_READ(type, addr, fieldU, fieldB) ((*((volatile type *) (addr))).fieldB)
-
-static inline UVISOR_FORCEINLINE void uvisor_write32(uint32_t volatile *addr, uint32_t val)
+static UVISOR_FORCEINLINE void uvisor_write32(uint32_t volatile *addr, uint32_t val)
 {
     *(addr) = val;
 }
 
-static inline UVISOR_FORCEINLINE void uvisor_write16(uint16_t volatile *addr, uint16_t val)
+static UVISOR_FORCEINLINE void uvisor_write16(uint16_t volatile *addr, uint16_t val)
 {
     *(addr) = val;
 }
 
-static inline UVISOR_FORCEINLINE void uvisor_write8(uint8_t volatile *addr, uint8_t val)
+static UVISOR_FORCEINLINE void uvisor_write8(uint8_t volatile *addr, uint8_t val)
 {
     *(addr) = val;
 }
 
-static inline UVISOR_FORCEINLINE uint32_t uvisor_read32(uint32_t volatile *addr)
+static UVISOR_FORCEINLINE uint32_t uvisor_read32(uint32_t volatile *addr)
 {
     return *(addr);
 }
 
-static inline UVISOR_FORCEINLINE uint16_t uvisor_read16(uint16_t volatile *addr)
+static UVISOR_FORCEINLINE uint16_t uvisor_read16(uint16_t volatile *addr)
 {
     return *(addr);
 }
 
-static inline UVISOR_FORCEINLINE uint8_t uvisor_read8(uint8_t volatile *addr)
+static UVISOR_FORCEINLINE uint8_t uvisor_read8(uint8_t volatile *addr)
 {
     return *(addr);
 }
 
+/* The conditional statement will be optimised away since the compiler already
+ * knows the sizeof(type). */
+#define ADDRESS_READ(type, addr) \
+    (sizeof(type) == 4 ? uvisor_read32((volatile uint32_t * volatile) (addr)) : \
+     sizeof(type) == 2 ? uvisor_read16((volatile uint16_t * volatile) (addr)) : \
+     sizeof(type) == 1 ? uvisor_read8((volatile uint8_t * volatile) (addr)) : 0)
+
+/* The switch statement will be optimised away since the compiler already knows
+ * the sizeof_type. */
+static UVISOR_FORCEINLINE void __address_write(size_t sizeof_type, volatile uint32_t *addr, uint32_t val)
+{
+    switch(sizeof_type) {
+        case 4:
+            uvisor_write32((volatile uint32_t * volatile) addr, (uint32_t) val);
```

Comment:
This is due to an issue, reported on GCC. Without the `volatile` compiler optimizations for the read/write functions were preventing our inline assembly code to be compiled as desired. It was fixed in #172 and has #173 has been opened to keep track of it.

But now that I think about it, the change here was only introduced for consistency with the supported case — it's not really needed and it shouldn't be there. I will remove this in a new PR.